### ISSUE TITLE
[JENKINS-9688] Build runs on master without executor and get stuck

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#15836]
<p>In the ASF Jenkins, we do not allow builds to run on master. Thus, we have set the number of executors on master to zero. Even so, one build recently (<a href="https://builds.apache.org/hudson/job/Axis2/774/" class="external-link" target="_blank" rel="nofollow noopener">https://builds.apache.org/hudson/job/Axis2/774/</a>) got assigned to run on master. The build is configured with a label to run on one of the slaves, but somehow Jenkins assigned it incorrectly. <br/>
Running on master also meant it got stuck, presumably due to the lack of executors, without any way of stopping it.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=protocol7b">protocol7b</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-9688" target="_blank">Build runs on master without executor and get stuck</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly">stephenconnolly</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: core, locks-and-latches-plugin
<li><b>resolution</b>: Incomplete
<li><b>resolved</b>: 2014-10-12T16:27:15+00:00
<li><b>votes</b>: 9
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>In the ASF Jenkins, we do not allow builds to run on master. Thus, we have set the number of executors on master to zero. Even so, one build recently (<a href="https://builds.apache.org/hudson/job/Axis2/774/" class="external-link" target="_blank" rel="nofollow noopener">https://builds.apache.org/hudson/job/Axis2/774/</a>) got assigned to run on master. The build is configured with a label to run on one of the slaves, but somehow Jenkins assigned it incorrectly. 
Running on master also meant it got stuck, presumably due to the lack of executors, without any way of stopping it.</p></pre>
</details>
</details>
<details><summary><i>1 attachment</i></summary>

- [job_time_null.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/21903/job_time_null.png)
  > ![job_time_null.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/21903/job_time_null.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>veithen:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-9688?focusedId=157515&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-157515">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Some additional observations for this issue:</p>

<p>The Axis2 builds (there are in total 3 builds for 3 different branches) use locks. What we see is that sometimes two (or more) different builds (for different branches) where triggered at the same time by the completion of a common upstream build (Axiom trunk e.g.). Since the Axis2 builds use a common lock, one would expect that only one starts execution, while the others remain in the build queue. However, what happens is that sometimes, two builds start execution in parallel, with one waiting for the lock (i.e. instead of waiting in the build queue, it is assigned to an executor and waiting there).</p>

<p>Here is a screenshot that shows the problem (The three Axis2 builds all use the same lock):</p>

<p><a href="http://people.apache.org/~veithen/axis2-builds.png" class="external-link" target="_blank" rel="nofollow noopener">http://people.apache.org/~veithen/axis2-builds.png</a></p>

<p>While the build is waiting on the executor to acquire the lock, it is reported as running on the master. E.g. the axis2-1.6 #225 build showed:</p>

<p>"Started 42 min ago
Build is being executed for 42 min on master"</p>

<p>In that particular case, the axis2-1.6 build eventually completed successfully. However, I think (I'm 90% sure) that I saw another Axis2 build that was triggered after the axis2-1.6 #225, but that actually started execution before. That would mean that when blocked builds are assigned to executors, they are no longer executed in FIFO style.</p>

<p>As a conclusion, I think that in order to make progress on this issue, one should first concentrate on the issues that occur when using locks:</p>
<ul>
	<li>The fact that builds waiting for a lock (on an executor) may be reported as running on master makes it hard to debug things if something gets stuck.</li>
	<li>The fact that execution is not FIFO means that a build may appear to be stuck simply because there is a constant flow of other builds in the queue that use the same lock. (Unfortunately it is not possible to check if that is a valid explanation for the original issue reported in this JIRA)</li>
</ul>


<p>Note: these observations where made with the following build: Jenkins ver. 1.447-SNAPSHOT (private-01/01/2012 20:43 GMT-olamy)</p>
</pre>
</details>

</ul>
</details>

<p>Some additional observations for this issue:</p>

<p>The Axis2 builds (there are in total 3 builds for 3 different branches) use locks. What we see is that sometimes two (or more) different builds (for different branches) where triggered at the same time by the completion of a common upstream build (Axiom trunk e.g.). Since the Axis2 builds use a common lock, one would expect that only one starts execution, while the others remain in the build queue. However, what happens is that sometimes, two builds start execution in parallel, with one waiting for the lock (i.e. instead of waiting in the build queue, it is assigned to an executor and waiting there).</p>

<p>Here is a screenshot that shows the problem (The three Axis2 builds all use the same lock):</p>

<p><a href="http://people.apache.org/~veithen/axis2-builds.png" class="external-link" target="_blank" rel="nofollow noopener">http://people.apache.org/~veithen/axis2-builds.png</a></p>

<p>While the build is waiting on the executor to acquire the lock, it is reported as running on the master. E.g. the axis2-1.6 #225 build showed:</p>

<p>"Started 42 min ago<br/>
Build is being executed for 42 min on master"</p>

<p>In that particular case, the axis2-1.6 build eventually completed successfully. However, I think (I'm 90% sure) that I saw another Axis2 build that was triggered after the axis2-1.6 #225, but that actually started execution before. That would mean that when blocked builds are assigned to executors, they are no longer executed in FIFO style.</p>

<p>As a conclusion, I think that in order to make progress on this issue, one should first concentrate on the issues that occur when using locks:</p>
<ul>
	<li>The fact that builds waiting for a lock (on an executor) may be reported as running on master makes it hard to debug things if something gets stuck.</li>
	<li>The fact that execution is not FIFO means that a build may appear to be stuck simply because there is a constant flow of other builds in the queue that use the same lock. (Unfortunately it is not possible to check if that is a valid explanation for the original issue reported in this JIRA)</li>
</ul>


<p>Note: these observations where made with the following build: Jenkins ver. 1.447-SNAPSHOT (private-01/01/2012 20:43 GMT-olamy)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>veithen:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-9688?focusedId=160122&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-160122">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Still occurs (with identical symptoms) on Jenkins 1.454.</p>
</pre>
</details>

</ul>
</details>

<p>Still occurs (with identical symptoms) on Jenkins 1.454.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/wbauer.png" title="wbauer's avatar" /> wbauer:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-9688?focusedId=160980&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-160980">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I see the same issue sporadically (Jenkins 1.454 currently).
The build seems to be stuck in the assignment phase and no thread or sockets refers to it at all.
The build is not visible in any queue or listing except on the job itself.
The build shows:
   "Started 2 hr 1 min ago
   Build is being executed for null on master"
.
This build is not killable meaning that I have to restart the master to get rid of the build...</p>
</pre>
</details>

</ul>
</details>

<p>I see the same issue sporadically (Jenkins 1.454 currently).<br/>
The build seems to be stuck in the assignment phase and no thread or sockets refers to it at all.<br/>
The build is not visible in any queue or listing except on the job itself.<br/>
The build shows:<br/>
   "Started 2 hr 1 min ago<br/>
   Build is being executed for null on master"<br/>
.<br/>
This build is not killable meaning that I have to restart the master to get rid of the build...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>josesa:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-9688?focusedId=163284&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-163284">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The problem still occurs in 1.458 see:
<span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/21903/21903_job_time_null.png" style="border: 0px solid black" /></span></p>

<p>We have multiple jobs using the same lock representing a test environment, but without any quiet period. These jobs are not meant to run on master, but on a build slave, but in the build status it shows:
"Started 1 day 14 hr ago
Build is being executed for null on master"</p>

<p>Maybe activating the default quiet period (30 seconds) will help in this situation.</p>
</pre>
</details>

</ul>
</details>

<p>The problem still occurs in 1.458 see:<br/>
<span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-core-cli/refs/heads/main/attachments/21903/job_time_null.png" style="border: 0px solid black" /></span></p>

<p>We have multiple jobs using the same lock representing a test environment, but without any quiet period. These jobs are not meant to run on master, but on a build slave, but in the build status it shows:<br/>
"Started 1 day 14 hr ago<br/>
Build is being executed for null on master"</p>

<p>Maybe activating the default quiet period (30 seconds) will help in this situation.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>josesa:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-9688?focusedId=163422&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-163422">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Even with delays it got stuck (maybe because I used exact same quiet period). Now I've configured different quiet periods to avoid this.</p>
</pre>
</details>

</ul>
</details>

<p>Even with delays it got stuck (maybe because I used exact same quiet period). Now I've configured different quiet periods to avoid this.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>josesa:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-9688?focusedId=167890&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-167890">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>After several similar situations in the past weeks like this one where concurrent jobs using the same lock I think the problem may be mostly between chair and keyboard, but also due to lack of informative messages.</p>

<p>Lets assume 3 jobs A, B and C using the same lock, the following timeline occurs:</p>
<ol>
	<li>A is already running (so it has the lock) and we start B and C.</li>
	<li>B and C show in queue stating they are being blocked by A</li>
	<li>A finishes</li>
	<li>B and C already waited the 'delay' so they are ready to run, so they do run at the same time</li>
	<li>B gets the lock so it states it in the log is activelly running</li>
	<li>C on the other hand didn't get the lock, but is running anyway and showing no output that it is actively waiting for the lock</li>
	<li>Users panic when they see a job running for 2 days without any output in the log and cancel the C</li>
	<li>C get's inconsistent state and starts showing "null" instead of real dates</li>
	<li>Jenkins needs to be restarted to get rid of C because there is no way to kill it</li>
</ol>


<p>Also we observed that if we don't Cancel C eventually it starts executing normally when B finishes and showing it has acquired the lock in the log file.</p>

<p>Bottom line the problem is lack of informative messages from the plugin to provide the proper feedback to users so they don't panic.</p>

<p>Workaround: Revoke users permissions to Cancel builds that are using locks.</p>
</pre>
</details>

</ul>
</details>

<p>After several similar situations in the past weeks like this one where concurrent jobs using the same lock I think the problem may be mostly between chair and keyboard, but also due to lack of informative messages.</p>

<p>Lets assume 3 jobs A, B and C using the same lock, the following timeline occurs:</p>
<ol>
	<li>A is already running (so it has the lock) and we start B and C.</li>
	<li>B and C show in queue stating they are being blocked by A</li>
	<li>A finishes</li>
	<li>B and C already waited the 'delay' so they are ready to run, so they do run at the same time</li>
	<li>B gets the lock so it states it in the log is activelly running</li>
	<li>C on the other hand didn't get the lock, but is running anyway and showing no output that it is actively waiting for the lock</li>
	<li>Users panic when they see a job running for 2 days without any output in the log and cancel the C</li>
	<li>C get's inconsistent state and starts showing "null" instead of real dates</li>
	<li>Jenkins needs to be restarted to get rid of C because there is no way to kill it</li>
</ol>


<p>Also we observed that if we don't Cancel C eventually it starts executing normally when B finishes and showing it has acquired the lock in the log file.</p>

<p>Bottom line the problem is lack of informative messages from the plugin to provide the proper feedback to users so they don't panic.</p>

<p>Workaround: Revoke users permissions to Cancel builds that are using locks.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dimitkoto:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-9688?focusedId=197617&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-197617">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I still have the same issue, using Jenkins ver. 1.552 on 64bit Linux RHEL, and only Restart of Jenkins is able to resolve these zombie jobs staying there forever and preventing new builds from starting.</p>
</pre>
</details>

</ul>
</details>

<p>I still have the same issue, using Jenkins ver. 1.552 on 64bit Linux RHEL, and only Restart of Jenkins is able to resolve these zombie jobs staying there forever and preventing new builds from starting.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-9688?focusedId=213701&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213701">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Too old and too little detail in the issue report to help further investigation, so resolving as incomplete.</p>

<p>Please file a new issue if this happens with a recent version of Jenkins (no older than ten weeks or so), and provide relevant information: Installed plugins, job configuration, how this can be reproduced, etc. See also the wiki for what other information may be relevant:
<a href="https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue</a></p>
</pre>
</details>

</ul>
</details>

<p>Too old and too little detail in the issue report to help further investigation, so resolving as incomplete.</p>

<p>Please file a new issue if this happens with a recent version of Jenkins (no older than ten weeks or so), and provide relevant information: Installed plugins, job configuration, how this can be reproduced, etc. See also the wiki for what other information may be relevant:<br/>
<a href="https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/How+to+report+an+issue</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>guptapranav01:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-9688?focusedId=238486&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-238486">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi,</p>

<p>whats the best way to run couple of jenkins jobs using groovy script. I don't find any good script for this. Any kind of help is greatly appreciated</p>
</pre>
</details>

</ul>
</details>

<p>Hi,</p>

<p>whats the best way to run couple of jenkins jobs using groovy script. I don't find any good script for this. Any kind of help is greatly appreciated</p>
